### PR TITLE
Add version and checksum to load_info struct

### PR DESCRIPTION
**Pull request by ppannuto**
Version numbering should allow us to catch issues from changes to the load_info
struct in the future. The checksum is currently just a basic XOR of all the fields
in the load_info struct.

Run-tested on storm, compile-tested on other platforms.

### CODE REVIEW

**Comment by ppannuto** [comment]
Is there a good way to tell rustfmt it's wrong, b/c this is strictly worse IMHO:

```
-    let checksum = load_info.version
-        ^ load_info.total_size
-        ^ load_info.entry_offset
-        ^ load_info.rel_data_offset
-        ^ load_info.rel_data_size
-        ^ load_info.text_offset
-        ^ load_info.text_size
-        ^ load_info.got_offset
-        ^ load_info.got_size
-        ^ load_info.data_offset
-        ^ load_info.data_size
-        ^ load_info.bss_mem_offset
-        ^ load_info.bss_size
-        ^ load_info.pkg_name_offset
-        ^ load_info.pkg_name_size;
+    let checksum =
+        load_info.version ^ load_info.total_size ^ load_info.entry_offset ^
+        load_info.rel_data_offset ^ load_info.rel_data_size ^ load_info.text_offset ^
+        load_info.text_size ^ load_info.got_offset ^
+        load_info.got_size ^ load_info.data_offset ^ load_info.data_size ^
+        load_info.bss_mem_offset ^ load_info.bss_size ^ load_info.pkg_name_offset ^
+        load_info.pkg_name_size;
```

**Comment by ppannuto** [comment]
Eh, I got over it and pushed the fmt friendly version

**Comment by alevy** [comment]
@ppannuto I wonder if just placing the `^` with the previous line would satisfy it?

**Comment by ppannuto** [comment]
Nope. It'll make for an uglier diff if/when that ever changes, but the benefits of consistency way outsize that nit. It's fine how it is.

**Comment by alevy** [comment]
Not sure why it's blocked, but feel free to merge when you're ready

**Comment by ppannuto** [comment]
I blocked it because I want to make a few more changes to the load_info struct. I figured once we introduce versioning of the header I don't want things running out of control and I'll just keep adding to this PR.

My current task queue:
   - Get a cxx_hello example using `std::string` working (prove cpp libs work)
      - Including string puts us over 8K of memory
          - Modify process compilation / loading to put process size in app header and dynamically size apps on loading in Tock <-- mostly implemented, hopefully will finish today

File: kernel/src/process.rs [hunk 1]
```diff
@@ -91,6 +92,34 @@ struct LoadInfo {
...
+/// not perform sanity or security checking on the structure
+unsafe fn parse_and_validate_load_info(address: *const u8) -> Option<&'static LoadInfo> {
+    let load_info = &*(address as *const LoadInfo);
+
+    if load_info.version != 1 {
+        return None;
+    }
+
+    let checksum =
+        load_info.version ^ load_info.total_size ^ load_info.entry_offset ^
```
Comment (brghena) [inline kernel/src/process.rs]: Do you not need to xor with 1 first here (as you did in `userland/tools/elf2tbf/src/main.rs` line 168)?

Comment (ppannuto) [inline kernel/src/process.rs]: No, the `1` from before is the `load_info.version` here.

**Comment by bradjc** [comment]
This doesn't seem to compile on hail and throws warning on the other platforms.

**Comment by ppannuto** [comment]
I'll fix hail later tonight. We should add the boards matrix to Travis.

The unused variable warnings are intentional. They're for Branden's stuff
which will use them

**Comment by bradjc** [comment]
Is the checksum currently checked anywhere?

**Comment by bradjc** [comment]
It would be great to document how to make apps work that need more stack/heap space in probably doc/Userland.md and the userland readme.

**Comment by ppannuto** [comment]
Yes the checksum is verified. In the parse_load_info function (or something
like that).

Will add doc

**Comment by bradjc** [comment]
I realized later that where the checksum is verified github helpfully decided that the diff was too long and just hid it. Which is strange, because it seems like the thing you would most want to see is the file with the biggest diff (assuming it's not a new file).

File: kernel/src/process.rs [hunk 2]
```diff
@@ -244,73 +301,115 @@ impl<'a> Process<'a> {
...
+                     app_flash_address,
+                     remaining_app_memory,
+                     remaining_app_memory_size) {
+                let stack_len = align8!(load_info.min_stack_len);
+                let app_heap_len = align8!(load_info.min_app_heap_len);
+                let kernel_heap_len = align8!(load_info.min_kernel_heap_len);
+
+                let app_slice_size =
+                    (load_result.data_len + stack_len + app_heap_len + kernel_heap_len) as usize;
+                // TODO round app_slice_size up to MPU unit?
```
Comment (bradjc) [inline kernel/src/process.rs]: Confirmed, this must be done.

Comment (alevy) [inline kernel/src/process.rs]: OK, good thing there is a function for that already.

**Pull request merged**